icechunk/src/session.rs

```diff
@@ -160,31 +161,79 @@ impl From<VirtualReferenceError> for SessionError {
+    // but only split (0,0) and split (1,1) are populated with data.
+    // A coord located in (1, 0) should return Err
+    // Since split_range need not form a regular grid, we must iterate through and find the first result.
+    // ManifestExtents in split_range MUST NOT overlap with each other. How do we ensure this?
+    // ndim must be the same
+    // debug_assert_eq!(coord.0.len(), split_range[0].len());
+    // FIXME: could optimize for unbounded single manifest
+    // Note: I don't think we can distinguish between out of bounds index for the array
+    //       and an index that is part of a split that hasn't been written yet.
+    iter.enumerate()
+        .find(|(_, e)| e.contains(coord.0.as_slice()))
+        .map(|(i, e)| (i, e.clone()))
+        .ok_or(
+            SessionErrorKind::InvalidIndexForSplitManifests { coords: coord.clone() }
+                .into(),
+        )
+}
+
 impl ManifestSplits {
-    // Returns the index of split_range that includes ChunkIndices
-    // This can be used at write time to split manifests based on the config
-    // and at read time to choose which manifest to query for chunk payload
-    pub fn which(&self, coord: &ChunkIndices) -> SessionResult<usize> {
-        // split_range[i] must bound ChunkIndices
-        // 0 <= return value <= split_range.len()
-        // it is possible that split_range does not include a coord. say we have 2x2 split grid
-        // but only split (0,0) and split (1,1) are populated with data.
-        // A coord located in (1, 0) should return Err
-        // Since split_range need not form a regular grid, we must iterate through and find the first result.
-        // ManifestExtents in split_range MUST NOT overlap with each other. How do we ensure this?
-        // ndim must be the same
-        // debug_assert_eq!(coord.0.len(), split_range[0].len());
-        // FIXME: could optimize for unbounded single manifest
-        // Note: I don't think we can distinguish between out of bounds index for the array
-        //       and an index that is part of a split that hasn't been written yet.
-        self.iter()
-            .enumerate()
-            .find(|(_, e)| e.contains(coord.0.as_slice()))
-            .map(|(i, _)| i)
-            .ok_or(
-                SessionErrorKind::InvalidIndexForSplitManifests { coords: coord.clone() }
-                    .into(),
-            )
+    pub fn which_extent_and_index(
+        &self,
+        coord: &ChunkIndices,
+    ) -> SessionResult<(usize, ManifestExtents)> {
+        which_extent_and_index(self.iter(), coord)
+    }
+
+    #[cfg(test)]
+    pub fn which_index(&self, coord: &ChunkIndices) -> SessionResult<usize> {
+        which_extent_and_index(self.iter(), coord).map(|(i, _)| i)
+    }
+}
+
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub enum Overlap {
+    Complete,
+    Partial,
+    None,
+}
+
+/// Important: this is not symmetric.
```

Comment:
maybe adding it with a different name to the `impl` would help make the asymmetry more clear?

Reply:
Now moved to `ManifestExtents.overlap_with(other: &ManifestExtents)`